docs/ADR-4-Support-only-for-mainnet-and-upcoming-eras.md

```diff
@@ -86,272 +94,67 @@ makeUnsignedTx era bc = ...
 ```
-### GADT flexibility
-
-The flexibility of GADTs also bites us as just by looking at the code it appears the deprecated code path is reachable, but it is not:
 
-```haskell
-eraSpecificLedgerTxBody
-  :: Era era
-  -> Ledger.TxBody (ToConstrainedEra era)
-  -> TxBodyContent BuildTx (AvailableErasToSbe era)
-  -> Either UnsignedTxError (Ledger.TxBody (ToConstrainedEra era))
-eraSpecificLedgerTxBody CurrentEra ledgerbody bc =
-  let propProcedures = txProposalProcedures bc
-      voteProcedures = txVotingProcedures bc
-      treasuryDonation = txTreasuryDonation bc
-      currentTresuryValue = txCurrentTreasuryValue bc
-  in return $
-       ledgerbody
-          & L.proposalProceduresTxBodyL .~ convProposalProcedures (maybe TxProposalProceduresNone unFeatured propProcedures)
-          & L.votingProceduresTxBodyL .~ convVotingProcedures (maybe TxVotingProceduresNone unFeatured voteProcedures)
-          & L.treasuryDonationTxBodyL .~ maybe (L.Coin 0) unFeatured treasuryDonation
-          & L.currentTreasuryValueTxBodyL .~ L.maybeToStrictMaybe (unFeatured <$> currentTresuryValue)
-
-eraSpecificLedgerTxBody UpcomingEra ledgerbody _bc = return ledgerbody
---  sbe <- maybe (Left $ error "eraSpecificLedgerTxBody: TODO") Right $ protocolVersionToSbe CurrentEra
---
---  setUpdateProposal <- first UnsignedTxError $ convTxUpdateProposal sbe (txUpdateProposal bc)
---
---  return $ ledgerbody  & L.updateTxBodyL .~ setUpdateProposal
-```
+This deprecation will happen only after the hardfork has been successful.
+There should be no need for consumers of the api to need the prior era's specific functionality.
+If users for some reason want this, we can direct them to use the ledger's api directly.
 
-Type error when called: 
+The next stage after deprecation period, should be **removal of the deprecated constructor**.
+The deprecation period should be long enough to give enough time for `cardano-api` consumers to update their codebase to the post-hardfork era.
 
-```haskell
-ghci> eraSpecificLedgerTxBody UpcomingEra undefined undefined
+## Downsides
 
-<interactive>:9:1: error: [GHC-64725]
-    • There is currently no planned upcoming era. Use CurrentEra instead.
-    • In a stmt of an interactive GHCi command: print it
-```
+`cardano-api` users will be limited up to only two eras exposed from `cardano-api`.
+This will force them to keep up-to-date their code after the hardfork. 
+This may turn out to be a disruptive process, but necessary to make the code using `cardano-api` healthy.
 
 # Approach
 
-The new api should be created adjacant to the existing one. We then slowly replace the use of the existing api in cardano-api, eventually deprecating the "old" api. 
+The new api should be created adjacant to the existing one.
+We then slowly replace the use of the existing api in cardano-api, eventually deprecating the "old" api. 
+
+With respect to cardano-cli, we should introduce top level `current-era` and `upcoming-era` commands.
```

Comment:
Thinking about this again we should just use the era. Why? These names would have the same problem we just ran into with `cardano-cli latest`.

Reply:
I think I agree with you. On the era boundary this will break most of the scripts relying on the `latest` being a particular era.